discretize/CylMesh.py

```diff
@@ -285,112 +1189,294 @@ def nodalLaplacian(self):
+        """
+        averaging operator of x-faces (radial) to cell centered vectors
+
+        :rtype: scipy.sparse.csr_matrix
+        :return: matrix that averages from faces to cell centered vectors
+        """
         if getattr(self, '_aveF2CCV', None) is None:
-            n = self.vnC
-            if self.isSymmetric:
-                avR = utils.av(n[0])[:, 1:]
-                avR[0, 0] = 1.
-                self._aveF2CCV = sp.block_diag((sp.kron(utils.speye(n[2]),
-                                                        avR),
-                                                sp.kron(utils.av(n[2]),
-                                                        utils.speye(n[0]))),
-                                               format="csr")
+            # n = self.vnC
+            if self.isSymmetric is True:
+                self._aveF2CCV = sp.block_diag(
+                    (self.aveFx2CC, self.aveFz2CC), format="csr"
+                )
             else:
-                raise NotImplementedError('wrapping in the averaging is not '
-                                          'yet implemented')
+                self._aveF2CCV = sp.block_diag(
+                    (self.aveFx2CC, self.aveFy2CC, self.aveFz2CC),
+                    format="csr"
+                )
         return self._aveF2CCV
 
+    ####################################################
+    # Deflation Matrices
+    ####################################################
+
+    def _deflationMatrix(self, location, withHanging=True, asOnes=False):
+        """
+        construct the deflation matrix to remove hanging edges / faces / nodes
+        from the operators
+        """
+        assert location in [
+            'N', 'F', 'Fx', 'Fy', 'Fz', 'E', 'Ex', 'Ey', 'Ez', 'CC'
+        ], (
+            'Location must be a grid location, not {}'.format(location)
+        )
+        if location == 'CC':
+            return speye(self.nC)
+
+        elif location in ['E', 'F']:
+            if self.isSymmetric:
+                if location == 'E':
+                    return self._deflationMatrix(
+                        'Ey', withHanging=withHanging, asOnes=asOnes
+                    )
+                elif location == 'F':
+                    return sp.block_diag([
+                        self._deflationMatrix(
+                            location+coord, withHanging=withHanging,
+                            asOnes=asOnes
+                        )
+                        for coord in ['x', 'z']
+                    ])
+            return sp.block_diag([
+                self._deflationMatrix(
+                    location+coord, withHanging=withHanging, asOnes=asOnes
+                )
+                for coord in ['x', 'y', 'z']
+            ])
+
+        R = speye(getattr(self, '_nt{}'.format(location)))
+        hanging = getattr(self, '_hanging{}'.format(location))
+        nothanging = ~getattr(self, '_ishanging{}'.format(location))
```

Comment:
Might be better to use this sort of logic on the grids. Rather than construct the deflated identity matrix.

Reply:
👍